README.md
# postcss-multi-value-display

[PostCSS] plugin to transform [CSS Display Module Level 3] multi-value syntax into legacy single-value equivalents (when available).

## Example

```js
const fs = require("fs")
const postcss = require("postcss")
const mvdisplay = require("postcss-multi-value-display")

let css = fs.readFileSync("input.css", "utf8")

let output = postcss()
  .use(mvdisplay)
  .process(css)
  .css
```

If `input.css` contains the following:

```css
.card-title {
  display: block flow;
}

.card-items {
  display: inline flex;
}
```

Then output will be:

```css
.card-title {
  display: block;
}

.card-items {
  display: inline-flex;
}
```

## Supported values

The following 'display' values have CSS level 2 single-value (legacy) equivalents and are supported by this plugin:

- `block flow` (equivalent to `block`)
- `block flow-root` (equivalent to `flow-root`)
- `inline flow` (equivalent to `inline`)
- `inline flow-root` (equivalent to `inline-block`)
- `run-in flow` (equivalent to `run-in`)
- `block flex` (equivalent to `flex`)
- `inline flex` (equivalent to `inline-flex`)
- `block grid` (equivalent to `grid`)
- `inline grid` (equivalent to `inline-grid`)
- `inline ruby` (equivalent to `ruby`)
- `block table` (equivalent to `table`)
- `inline table` (equivalent to `inline-table`)

Note that the two-value versions can be specified in either order (per the W3C spec), so e.g. both `inline flex` and `flex inline` will be transformed to `inline-flex`.

Other values from CSS Display Module Level 3 are **not supported**. This includes:

- `block ruby` as there is no single-value equivalent
- any of the multi-value `list-item` variants
- `contents` as this value introduces an entirely new semantic which cannot be emulated by transpiling

Also be aware that even among the single-value options which this plugin outputs, browser support is not universal. For example, `inline-block` is supported as far back as IE 8, but `flow-root` is unsupported in IE and in EdgeHTML-based (i.e. pre-Chromium) MS Edge versions.

## Options

This plugin can optionally be configured by passing in an options object, like this:

```js
const postcss = require("postcss")
const mvdisplay = require("postcss-multi-value-display")

const options = { ... }

postcss.use(mvdisplay(options)).process(...)
```

Allowed options are:

### `preserve` (default: false)

Keep two-value 'display' declarations but emit the appropriate single-value legacy syntax as a separate declaration above each. Browsers skip over declarations whose values they don't understand, so new browsers will use the latter line (overriding the former) while older ones will use the former and skip the latter.

This should generally never be necessary as the old values are completely equivalent to the new ones. In theory this means adding `{ preserve: true }` will not change any browsers' behavior and will only serve to bloat your output CSS.

## Warnings

This plugin will issue a warning if it encounters a `display` declaration with a multi-token value that it doesn't recognize. For example, if your CSS includes `display: block ruby;` somewhere and you run PostCSS from the CLI, you'd see a message like this:

```
6:3	⚠  Cannot transform 'display: block ruby' (no legacy equivalent exists). [postcss-multi-value-display]
```

## License

This repository is made available under the MIT license; see the included LICENSE file for details.

[PostCSS]: https://github.com/postcss/postcss
[CSS Display Module Level 3]: https://www.w3.org/TR/css-display-3


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2021-03-25

Initial release

[Unreleased]: https://github.com/jake-low/postcss-multi-value-display/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/jake-low/postcss-multi-value-display/releases/tag/v1.0.0


index.js
// See CSS Display Module Level 3
// https://www.w3.org/TR/css-display-3 (retrieved 2020-03-25)

const lookup = new Map([
  ["block flow", "block"],
  ["block flow-root", "flow-root"],
  ["inline flow", "inline"],
  ["inline flow-root", "inline-block"],
  ["run-in flow", "run-in"],
  // multi-value "list-item" variants excluded for simplicity;
  // only "block flow list-item" has a legacy equivalent ("list-item")
  // and they aren't order-independent like the values included here are.
  ["block flex", "flex"],
  ["inline flex", "inline-flex"],
  ["block grid", "grid"],
  ["inline grid", "inline-grid"],
  ["inline ruby", "ruby"],
  // ["block ruby"] has no single-value equivalent
  ["block table", "table"],
  ["inline table", "inline-table"],
])

function * permutations(arr) {
  // Uses Heap's method; see:  Sedgewick, R. “Permutation Generation Methods,”
  // Computing Surveys 9, 2 (June 1977), 141.
  const length = arr.length
  let c = Array(length).fill(0)
  let i = 1

  yield arr.slice()

  while (i < length) {
    if (c[i] < i) {
      let k = i % 2 ? c[i] : 0
      let temp = arr[i]
      arr[i] = arr[k]
      arr[k] = temp
      c[i] += 1
      i = 1
      yield arr.slice()
    } else {
      c[i] = 0
      i += 1
    }
  }
}

for (const [key, value] of [...lookup.entries()]) {
  for (const permutation of permutations(key.split(" "))) {
    lookup.set(permutation.join(" "), value)
  }
}

const defaults = {
  preserve: false
}

module.exports = (options = {}) => {
  options = { ...defaults, ...options }

  return {
    postcssPlugin: "postcss-multi-value-display",
    Declaration: {
      display: (decl, { result }) => {
        if (!/ /.test(decl.value)) return

        const replacement = lookup.get(decl.value)
        if (replacement === undefined) {
          decl.warn(result, `Cannot transform 'display: ${decl.value}' (no legacy equivalent exists).`, { word: decl.value })
          return
        }

        if (options.preserve) {
          decl.cloneBefore({ value: replacement })
        } else {
          decl.value = replacement
        }
      }
    }
  }
}

module.exports.postcss = true
